Support for use of Enums in MultiIndex

benediamond opened the pull request:
- [x] closes #21298
- [x] tests added / passed: added `test_use_enum_in_multiindex` in `tests/indexes/test_multi.py`, see diff
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The boolean flag in https://github.com/pandas-dev/pandas/blob/9f95f7dbffef7752175ca9ed918314cb6f0b9b18/pandas/core/arrays/categorical.py#L2526-L2528 incorrectly caused a `TypeError` to be raised in
https://github.com/pandas-dev/pandas/blob/9f95f7dbffef7752175ca9ed918314cb6f0b9b18/pandas/core/arrays/categorical.py#L352-L354
when an Enum was used in a MultiIndex. No tests fail when the flag is flipped back. This is fixed in this PR.

benediamond committed:
Use of Enums in MultiIndex now supported

Due to an unnecessary "orderedness" check in `arrays/categorical.py`'s method `_factorize_from_iterable`, MultiIndex's constructor inappropriately threw a `TypeError` when passed an Enum as one of its iterables. This has been fixed.

benediamond committed:
added a test for `MultiIndex` with `Enum`

benediamond committed:
Update test_multi.py

benediamond committed:
PEP8

benediamond commented:
failure is due to unavailability of `enum` in test environment...

gfyoung commented:
> CircleCI failed due to unavailability of module enum in test environment. I don't know how else to test this---should I simply remove the new test I added?

jreback commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -3307,3 +3307,40 @@ def test_duplicate_multiindex_labels(self):
+        # cannot actually be built yet.
+        from pandas.core.algorithms import take_1d
+
+        def _get_ilevel_values(index, level):
```
what is all this for?

simply construct the expected index directly

benediamond replied:
The thought here is that the expected index _cannot_ be constructed directly, due to the exact bug I'm trying to fix. Thus instead I create an expected "level", and verify that it matches the relevant level of `df`'s index. This is done by replicating a sub-step of `tm.assert_index_equal`.

jreback commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -3307,3 +3307,40 @@ def test_duplicate_multiindex_labels(self):
+            values = unique._shallow_copy(filled, name=index.names[level])
+            return values
+
+        from enum import Enum
```
all imports go at the top

jreback commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -3307,3 +3307,40 @@ def test_duplicate_multiindex_labels(self):
+            return values
+
+        from enum import Enum
+        MyEnum = Enum("MyEnum", "A B")
```
you would have to add ``enum34`` (the 2.7 backport) in our 2.7 builds

```
(pandas) bash-3.2$ grep python=2.7 ci/*.yaml
appveyor-27.yaml:  - python=2.7.*
circle-27-compat.yaml:  - python=2.7*
travis-27-locale.yaml:  - python=2.7
travis-27.yaml:  - python=2.7*
```

benediamond replied:
I'm sorry, does this mean simply importing `enum34` inline? Thank you.

benediamond replied:
ok, I've hopefully done this correctly.

benediamond committed:
Put imports on top

benediamond committed:
Update appveyor-27.yaml

benediamond committed:
Update circle-27-compat.yaml

benediamond committed:
Update travis-27-locale.yaml

benediamond committed:
Update travis-27.yaml

benediamond committed:
enum34 dependency

Added `enum34` dependency

benediamond committed:
add whatsnew

benediamond committed:
sync whatsnew with master

benediamond committed:
sync whatsnew with master

benediamond commented:
@jreback any idea why it's marking this `whatsnew` line as a conflict...?

benediamond committed:
attempting to resolve conflict

benediamond committed:
attempting to resolve conflicts in `whatsnew`

benediamond commented:
@gfyoung I'm stumped by this `whatsnew` conflict, any thoughts?

gfyoung commented:
> @gfyoung I'm stumped by this whatsnew conflict, any thoughts?

@benediamond : I would actually move your `whatsnew` addition to `0.24.0`.  That should take care of that 😄

gfyoung commented on doc/source/whatsnew/v0.23.1.txt:
```diff
@@ -88,6 +88,7 @@ Indexing
 - Bug in :meth:`MultiIndex.set_names` where error raised for a ``MultiIndex`` with ``nlevels == 1`` (:issue:`21149`)
 - Bug in :class:`IntervalIndex` constructors where creating an ``IntervalIndex`` from categorical data was not fully supported (:issue:`21243`, issue:`21253`)
 - Bug in :meth:`MultiIndex.sort_index` which was not guaranteed to sort correctly with ``level=1``; this was also causing data misalignment in particular :meth:`DataFrame.stack` operations (:issue:`20994`, :issue:`20945`, :issue:`21052`)
+- Bug in :func:`pandas.core.arrays.categorical._factorize_from_iterable` inappropriately caused ``TypeError`` to be raised when an ``Enum`` was used as a factor in a ``MultiIndex`` (:issue:`21298`)
```
> @gfyoung I'm stumped by this whatsnew conflict, any thoughts?

@benediamond : I would actually move your `whatsnew` addition to `0.24.0`.   That will suffice. 😄

benediamond replied:
I've gone ahead and just removed it---I'm not sure where it belongs, as it appeared that this PR (when it was passing) was slated for the next milestone. It was: "Bug in :func:`pandas.core.arrays.categorical._factorize_from_iterable` inappropriately caused ``TypeError`` to be raised when an ``Enum`` was used as a factor in a ``MultiIndex`` (:issue:`21298`)"

gfyoung replied:
@benediamond : I would put it in `0.24.0`, as I suggested above.  We will need a `whatsnew` for this for sure, and that would be the best place to put it.

benediamond replied:
@gfyoung Done. this one didn't create merge conflicts, thankfully

benediamond committed:
remove whatsnew

benediamond committed:
remove whatsnew

benediamond committed:
add to whatsnew 0.24

gfyoung commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -3307,3 +3307,39 @@ def test_duplicate_multiindex_labels(self):
+                labels, fill_value=unique._na_value
+            )
+            values = unique._shallow_copy(filled, name=index.names[level])
+            return values
```
Given that you use this function only once, I think this abstraction is unnecessary as @jreback alluded to.

benediamond replied:
Ok, amended. Thanks for your guidance.

gfyoung commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -3307,3 +3307,39 @@ def test_duplicate_multiindex_labels(self):
+        # GH 21298
+        # Allow use of Enums as one of the factors in a MultiIndex.
+        from enum import Enum
+        from pandas.core.algorithms import take_1d
```
By "top", @jreback meant "top of the file" 😄

benediamond replied:
Hah whoops. done

benediamond committed:
remove nested function from test

This nested function had been drawn directly from _within_ `pands.util.testing.assert_index_equal`. As there is no way to construct the target dataframe's (multi) index directly (hence this PR), I simply compare this index's _0th level_ to a manually constructed (simple) index. This comparison is given by a subroutine of `assert_index_equal` (which applies the subroutine to _each_ level).

benediamond committed:
move imports to top

gfyoung commented on pandas/core/arrays/categorical.py:
```diff
@@ -2523,7 +2523,7 @@ def _factorize_from_iterable(values):
         codes = values.codes
     else:
-        cat = Categorical(values, ordered=True)
+        cat = Categorical(values, ordered=False)
```
Interesting that this didn't cause any testing issues...

benediamond replied:
Yes, it's striking. It's possible that that`multi.py` is essentially the only caller of `categorical._factorize_from_iterable`, and that the calls that `multi` makes happen to be such that within `Categorical`'s constructor, control is always routed so that the only time that `ordered` is used is to incorrectly raise the `TypeError` on lines 352-354. That's my best-guess explanation, at least.

gfyoung reviewed (approved):
Well, this looks pretty good to me!

cc @jreback

benediamond commented:
@jreback bumping this.

jreback commented on ci/circle-27-compat.yaml:
```diff
@@ -22,6 +22,7 @@ dependencies:
   # universal
   - pytest
   - pytest-xdist
+  - enum34
```
can you remove from all but 1 of the 27 compat tests (the travis-27) one

jreback commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -3307,3 +3308,28 @@ def test_duplicate_multiindex_labels(self):
             ind.set_levels([['A', 'B', 'A', 'A', 'B'], [2, 1, 3, -2, 5]],
                            inplace=True)
+
+    def test_use_enum_in_multiindex(self):
```
can you use ``@td.skip_if_no('enum')`` here (then do the actual import inside), we have to guard against the tests being run on 2.7 w/o the compat library (by a user).

jreback commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -3307,3 +3308,28 @@ def test_duplicate_multiindex_labels(self):
+        filled = take_1d(unique.values, labels, fill_value=unique._na_value)
+        df_index_0 = unique._shallow_copy(filled, name=df.columns.names[0])
+        exp_index_0 = pd.Index([MyEnum.A, MyEnum.A, MyEnum.B, MyEnum.B],
+                               dtype='object')
```
this is pretty complicated, why cannot you just directly construct the Index here?

jreback reviewed (changes requested):
pls add a whatsnew note for 0.24

benediamond committed:
jreback fixes

benediamond commented:
for whatsnew, see: https://github.com/benediamond/pandas/blob/4cd2dff96775a828687acd7c65919ec135824cf4/doc/source/whatsnew/v0.24.0.txt#L132-L135

benediamond commented:
> can you use @td.skip_if_no('enum') here (then do the actual import inside), we have to guard against the tests being run on 2.7 w/o the compat library (by a user).

@jreback implemented here:
```python
    @td.skip_if_no('enum')
    def test_use_enum_in_multiindex(self):
        # GH 21298
        # Allow use of Enums as one of the factors in a MultiIndex.
        from enum import Enum
        MyEnum = Enum("MyEnum", "A B")
        ...
```
but yields:
```
==================================== ERRORS ====================================
_____________ ERROR collecting pandas/tests/indexes/test_multi.py ______________
pandas/tests/indexes/test_multi.py:30: in <module>
    class TestMultiIndex(Base):
pandas/tests/indexes/test_multi.py:3310: in TestMultiIndex
    @td.skip_if_no('enum')
E   NameError: name 'td' is not defined
------- generated xml file: /tmp/circle-junit.2ZpyFLd/reports/junit.xml --------
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
===================== 3 skipped, 1 error in 43.23 seconds ======================
```
not sure what's going on 😩

TomAugspurger commented:
I think you need to `import pandas.util._test_decorators as td` at the top of `test_multi.py`

benediamond committed:
import pandas.util._test_decorators as td

jreback commented:
closing as stale, if you'd like to continue, pls ping.

benediamond commented:
@jreback, my understanding was that this commit was ready, but had been waiting for your approval. I found it hard to get responses from you. if you're ready to go, i'm ready to play my part. it should be almost there.

TomAugspurger commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -132,7 +132,7 @@ Strings
 ^^^^^^^^
 
--
+- Bug in :func:`pandas.core.arrays.categorical._factorize_from_iterable` inappropriately caused ``TypeError`` to be raised when an ``Enum`` was used as a factor in a ``MultiIndex`` (:issue:`21298`)
```
Release notes shouldn't reference private methods.

You should be able to reference Enum with :class:`enum.Enum`, and :class:`MultiIndex`.

Commit:
rebase

Commit:
added test

benediamond commented:
@jreback and others: looks like this issue has _now been fixed_ by PR #22072, see also issue #15457. I have gone ahead and preserved the test I wrote in my rebase, in case you'd like to merge it in---if not no worries and this (as well as issue #21298) can be closed! (note that issue #15457 was a somewhat different problem). thanks for everyone's time and patience.

Commit:
typo: remove 'self'

WillAyd commented:
@benediamond closing per your latest comment. Ping if we need to reopen

scratchmex commented:
I want to point out that when merging dataframes with enums as multi-index it throws
`RuntimeWarning: The values in the array are unorderable. Pass `sort=False` to suppress this warning.`
so I think the support for enums is not yet completed

jreback commented:
you could open a separate issue with a reproducible example

but we have no official support for enums and would need a champion in the community
